### ISSUE TITLE
Integration with Mockito

### ISSUE THREAD

**Issue by petergphillips** [greghaskins/spectrum#15]
Hi Greg,

I have been having at Spectrum and really like the way the tests get grouped together.  I wondered if there was any support for mockito annotations, or if you knew a way to get them working.

I have managed to get a Mockito test by explicitly constructing the mocks and then adding a manual call to validateMockitoUsage and reset in the afterEach, it just feels quite hard work.

Normally I would add in a Mockito JUnit rule, annotate each mock and then construct the service in a before method.  Having to manually construct the mocks is not too bad, but It would be great if there was some way to trigger the validate and reset at the end of each test.

Thanks,

Peter.

**Comment by greghaskins** [comment]
Interesting concept. I hadn't even considered how using Mockito might feel with Spectrum. I might put some thought into it if I can get some spare cycles. 

Do you have some sample code I could check out using the technique you described?

**Comment by petergphillips** [comment]
I've taken one of your tests using Mockito and converted it into an example - please see https://github.com/petergphillips/spectrum/commit/bb85dd87968f14d896688f33178bfd63e3aa02eb.  Obviously not a good way to test Spectrum, but should give you an idea of how I had to integrate it...

**Comment by ashleyfrieze** [comment]
#49 should fix this. Once you can use rules to interact with objects in the test class's instance a few opportunities open up.

**Comment by ashleyfrieze** [comment]
I've made a hacked version of rules in Spectrum - see the spec file, rather than the implementation:

https://github.com/ashleyfrieze/spectrum/blob/scratch-implementation-rules/src/test/java/specs/MockitoSpec.java

```java
@RunWith(Spectrum.class)
public class MockitoSpec {
	@Mock
	private SomeInterface mockInterface;

	@InjectMocks
	private SomeClass objectUnderTest;

	@Rule
	public MockitoRule rule = MockitoJUnit.rule();

	{
		describe("A suite which needs mockito", () -> {
			beforeEach(() -> {
				given(mockInterface.getInput()).willReturn("Hello world");
			});

			it("can use the mocks", () -> {
				assertThat(objectUnderTest.getResult(), is("Hello world"));
			});

			it("can use the mocks again", () -> {
				assertThat(objectUnderTest.getResult(), is("Hello world"));
			});

			it("gets a fresh mock each time", () -> {
				verify(mockInterface, never()).getInput();
			});

		});
	}
}
```

Assume the supporting mocked interfaces around. This, for me, reads perfectly logically. We presume a fresh set of mocks for each `it` and we get one. What's not to like. Note - this spec passes the tests.

**Comment by greghaskins** [comment]
It works, but I don't think it's the best path for Spectrum. Here's why:

1. It just looks weird. Fields with annotations may make sense for people coming from a traditional JUnit experience, but that is just funky for someone who already knows RSpec or Jasmine and is looking for a comparable tool in Java-land.
2. It requires you to declare mocks globally for all suites and specs in the file. If I only need a mock for one particular nested context, I couldn't declare it right there. I'd have to declare it up at the top, and it would be around for all the other specs that don't need it. One of the key benefits (to me) of spec-style test runners is that they make it easier to write specific, focused tests. The tendency of many traditional JUnit tests I've seen is to add lots and lots of shared fields, many of which are only needed by a subset of the test cases. Nested `describe` blocks help you focus in.
3. I have a love/hate relationship with reflection and "magic" in general, especially with test suites. Believe it or not, the very first POC of Spectrum was all based on annotations and run-time reflection. It was pretty magical. And a real pain to debug. Add that to my nightmares chasing crazy stacktraces and `InvocationTargetException`s on production apps, I ended up on the current implementation of Spectrum which has basically zero reflection except a single `newInstance()` call. It's an opinionated stance, though.
4. I don't actually think it's that hard to write Mockito integration in a Spectrum-friendly, functional way.

To re-work your example, something like this would be pretty cool:

```java
@RunWith(Spectrum.class)
public class MockitoSpec {
  {
    describe("A suite which needs mockito", () -> {

      Supplier<SomeInterface> mockInterface = let(() -> mock(SomeInterface.class));
      Supplier<AnotherInterface> mockAnother = let(() -> mock(AnotherInterface.class));
      
      Supplier<SomeClass> objectUnderTest = injectMocks(SomeClass.class, mockInterface, mockAnother);
      
      beforeEach(() -> {
        given(mockInterface.get().getInput()).willReturn("Hello world");
      });

      it("can use the mocks", () -> {
        assertThat(objectUnderTest.get().getResult(), is("Hello world"));
      });

      it("can use the mocks again", () -> {
        assertThat(objectUnderTest.get().getResult(), is("Hello world"));
      });

      it("gets a fresh mock each time", () -> {
        verify(mockInterface.get(), never()).getInput();
      });

    });
  }
}
```

We could get fancy with the API (like `inject(SomeClass.class).withMock(mockThing)` or something), but that could be the overall flow.

The key differences:
- 👍 Functional style matching the rest of Spectrum and RSpec/Jasmine
- 👍 You could use `injectMocks` at any nesting level
- 👎 You'd have to use `.get()` instead of the implicit `this` (same as all your other local variables)
- 👍 👎 You'd have to explicitly pass your mocks to `injectMocks`, rather than the framework magically finding them for you. (I tend to prefer constructor injection anyway, which already works with Spectrum.)
- 👎 We'd have to re-implement some of the functionality built into the Mockito library, and we'd have to use some constructor/property reflection if we wanted it to be equally magical as `@InjectMocks` (assuming there isn't much we could re-use from Mockito's implementation).
- 👍 Our `injectMocks` would not need to do any crazy run-time annotation processing or as much validation of the test class itself. Looking through Mockito's code, there are tons of places where they have to make sure everything is exactly right because they get no help from the compiler.

**Comment by ashleyfrieze** [comment]
> :-1: We'd have to re-implement some of the functionality built into the Mockito library

We'd have to implement plugins for every single library which can currently plug into JUnit via `@Rule`

Here's a proposed compromise for "normal for Spectrum".

```java
class SomeRuleMixinClass {
    @Rule
    public MockitoJUnitRule useMockito = new MockitoJUnitRule();

    @Mock
    SomeInterface someMock;
}

@RunWith(Spectrum.class)
class TestClass {{
    describe("some feature", () -> {
        Supplier<SomeRuleMixinClass> mocks = applyRules(SomeRuleMixinClass.class);

        it("depends on the rules", () -> {
            when(mocks.get().someMock). ... // etc
        });
    });
}}
```

The above example is pointless for a single mock, it would be easier not to use it. But shows how a mixin would work. The mix-in class could be separate or made local to the test. Making them separate may be a massive boost to test modularisation!

The reason why the mixin class, and rule-controlled object is important is that JUnit assume a per-test instance of the test class, in which the rules live, where Spectrum assumes no special instance of the class which defines Spectrum tests.

I think that the mechanism that makes the above work (I think I can make it) would allow a more terse form of rule application using the local test class if people (and rules) want/allow it. But the mechanism would be pure enough for Spectrum's design intent of anything anywhere.

**Comment by ashleyfrieze** [comment]
@greghaskins - in my humble opinion, it seems that there's a potentially singular solution for #15 #49 and #34 - is it worth trying to coalesce them?

**Comment by greghaskins** [comment]
That looks really great, @ashleyfrieze. Really excellent idea. You could theoretically even nest that Mixin class inside your TestClass (with `static`, so TestClass.Mixin`) and then you even get it declared right where you need it. It seems to be the best of both worlds.

Based on your feedback, I realized that I was undervaluing the ability to leverage existing JUnit rules and infrastructure, which has the nice benefit of not needing to re-invent the world every time someone needs a new integration. For the most valuable things, we might create our own functional-style integration, but for others we should be able to leverage the existing `Rule` ecosystem.

Let's move this discussion over to #49 since that seems the be the generic approach we're chasing.

**Comment by ashleyfrieze** [comment]
:+1:

**Comment by ashleyfrieze** [comment]
> You could theoretically even nest that Mixin class inside your TestClass

Valid Java?

```java
@RunWith(Spectrum.class)
class TestClass {{
    describe("some feature", () -> {
        class SomeRuleMixinClass {
            @Rule
            public MockitoJUnitRule useMockito = new MockitoJUnitRule();

            @Mock
            SomeInterface someMock;
        }
        Supplier<SomeRuleMixinClass> mocks = applyRules(SomeRuleMixinClass.class);

        it("depends on the rules", () -> {
            when(mocks.get().someMock). ... // etc
        });
    });
}}
```

**Comment by greghaskins** [comment]
I think it's valid Java, but there might still be other gotchas like the implicit `this` of the containing scope. At least as a first cut, we can do a standard static nested class (or a completely separate class file) and not try to get too fancy yet.

**Comment by ashleyfrieze** [comment]
@petergphillips - please review #56

**Comment by petergphillips** [comment]
@ashleyfrieze - I've had a play around and looks really good.

It's exactly what I was after in terms of reusing existing `@Rule`, `@Mock` and even `@Captor` annotations.

I've tried converting one of my existing Mockito tests with the fork and it runs fine.  I then moved onto a Dropwizard test using mock-server (http://www.mock-server.com/) rules and it fell over during the rule initialisation.  So in my case I have the following code
```java
  @ClassRule
  public static final DropwizardAppRule<MyAppConfiguration> APP =
      new DropwizardAppRule<>(MyAppApplication.class,
                              ResourceHelpers.resourceFilePath("my-app-test-config.yaml"));

  @Rule
  public MockServerRule mockServerRule = new MockServerRule(this, Integer.parseInt(APP.getConfiguration()
                                                                                      .getMyUri()
                                                                                      .split(":")[2]));

```
The `@Rule` depends on the `@ClassRule` being run first and initialised.  I get a NullPointerException for the call to `getMyUri()`, indicating that `getConfiguration()` is returning null.

**Comment by ashleyfrieze** [comment]
Hi @petergphillips - this is great feedback. Thank you.

I think you've hit on a limitation of Spectrum in terms of this rule stuff. In your example, you don't want the TestObject - i.e. the instance of your class with the method rule in it - to be created before the class rule has run. If you make the test class the direct host of the rules, then Spectrum has already built the test object before it works out what rules to apply to it. This is because Spectrum is not itself running JUnit rules, but instead wiring them, in the right order of precedence, in with the functional rules specified using things like `aroundAll` or `beforeEach`.

Not to worry. This limitation was expected, and we provided two ways of running JUnit rules. Unless you tell me otherwise, in which I'll develop a fix, I'm going to guess you just dropped all your rules directly into the class with the spectrum tests in - e.g.

```java
@RunWith(Spectrum.class)
public class MyTest {
   @Rule
   public SomeRule ...; //etc

   {
      describe("sometest suite", () -> {
          it("has a test", ()-> {});
      });
   }
}
```

In the above, the test object is created with all its static and instance members and you can't do what you want. If we take your DropWizard code and put it into a mixin class:

```java
public DropWizardMixin {
  @ClassRule
  public static final DropwizardAppRule<MyAppConfiguration> APP =
      new DropwizardAppRule<>(MyAppApplication.class,
                              ResourceHelpers.resourceFilePath("my-app-test-config.yaml"));

  @Rule
  public MockServerRule mockServerRule = new MockServerRule(this, Integer.parseInt(APP.getConfiguration()
                                                                                      .getMyUri()
                                                                                      .split(":")[2]));
}
```

The mix-in can be used with the `applyRules` keyword and I _think_ it will work because on mix-ins the class rules run before the object is created.

```java
@RunWith(Spectrum.class)
public class MyTest {
   {
      Supplier<DropWizardMixin> referenceToCurrentMixin = applyRules(DropWizardMixin .class);
      describe("sometest suite", () -> {
          it("has a test", ()-> {})

          it("can access the mixin", () -> {
              referenceToCurrentMixin .get().mockServerRule.doSomething();
          });
      });
   }
}
```

If you've not already done so, would you mind road testing this approach?

**Comment by greghaskins** [comment]
Resolved by #56. Closing.
